### PR TITLE
Interactive mode with command or script

### PR DESCRIPTION
**Pull request by laerus**
this patch adds functionality similar to the python repl
when forcing interactive mode with command or script :
`xonsh -ic '...'`
`xonsh -i script.xsh`

### CODE REVIEW

File: xonsh/main.py [hunk 1]
```diff
@@ -386,6 +372,20 @@ def func_sig_ttin_ttou(n, f):
             code = sys.stdin.read()
             run_code_with_cache(code, shell.execer, glb=shell.ctx, loc=None,
                                 mode='exec')
+        if args.force_interactive:
```
Comment (scopatz) [inline xonsh/main.py]: Shouldn't this be an `elif`?

Comment (scopatz) [inline xonsh/main.py]: Also, why move away from `args.mode`?  Shouldn't args.force_interactive just set `args.mode` to be interactive in premain?

**Comment by scopatz** [comment]
Thanks for putting this in @laerus! Just a couple of comments in-line. Also, it looks like there are some test failures.

File: xonsh/main.py (hunk 1 above)
Comment (laerus) [inline xonsh/main.py]: we want xonsh to run the command or script before going into interactive,
modes are exclusive, we need new modes that correspond to 'interactive with script' and 'interactive with command'. With command/script argument present premain just sets mode accordingly so we can't check against mode for interactive. I was checking both against `args.force_interactive` and `args.mode` but it seemed redundant so i removed it.

File: tests/conftest.py [hunk 2]
```diff
@@ -72,40 +73,8 @@ def xonsh_builtins(xonsh_events):
...
-    if hasattr(builtins, '__xonsh_history__'):
-        del builtins.__xonsh_history__
-    del builtins.__xonsh_enter_macro__
-    del builtins.evalx
-    del builtins.execx
-    del builtins.compilex
-    del builtins.aliases
-    del builtins.events
+    for attr in set(dir(builtins)) - old_builtins:
+        delattr(builtins, attr)
```
Comment (scopatz) [inline tests/conftest.py]: Nice fix :)

File: xonsh/main.py (hunk 1 above)
Comment (scopatz) [inline xonsh/main.py]: This seems like you could end up running a script from a file and *then* entering a command loop. That shouldn't happen, even if it is in interactive mode. The command loop should be entered only if we are in interactive mode and we are not running a script  or a command (`-c`).

Comment (scopatz) [inline xonsh/main.py]: Yeah, I just tried it.

```console
scopatz@artemis:~$ xonsh -i -c "echo hi"   # in bash
hi
scopatz@artemis ~ $  exit # in xonsh 
scopatz@artemis:~$  # back in bash
```

We should never enter into xonsh fully. Take the inverse, for example,

```console
scopatz@artemis ~/xonsh/news nullsub $ bash -i -c "echo hello"  # in xonsh
bash: cannot set terminal process group (-1): Inappropriate ioctl for device
bash: no job control in this shell
hello
scopatz@artemis ~/xonsh/news interactive $ # still in xonsh
```

Does this make sense?

**Comment by scopatz** [comment]
Hi @laerus, I am seeing the following issue when I try to launch from bash:

```console
scopatz@artemis:~$ XONSH_DEBUG=1 xonsh
Traceback (most recent call last):
  File "/home/scopatz/.local/lib/python3.5/site-packages/xonsh/main.py", line 337, in main
    return main_xonsh(args)
  File "/home/scopatz/.local/lib/python3.5/site-packages/xonsh/main.py", line 354, in main_xonsh
    if args.mode == XonshMode.single_command:
AttributeError: 'Namespace' object has no attribute 'mode'
Xonsh encountered an issue during launch
Failback to /bin/sh
```

File: xonsh/main.py (hunk 1 above)
Comment (laerus) [inline xonsh/main.py]: yeah, that is how the python interactive option works, you can run a script and then enter the REPL in the same context. It seems like in bash it doesnt make any difference if you use the `-i` flag. Isnt this feature something we want in xonsh?

Comment (scopatz) [inline xonsh/main.py]: I don't think so. I think we want to match the behaviour of bash and other shells. Text editors and other tools rely on the these command line options to work in a certain way and we want to have comparability with those tools. Changing `-i` from non-blocking to blocking would probably break a bunch of these things.

**Comment by scopatz** [comment]
Hi @laerus I am still in favor of the test refactor here, if you are interested in cleaning that up.

**Comment by laerus** [comment]
Hey @scopatz, this started out as a general xonsh.main and tests refactor when i noticed the difference in interactive mode with the python REPL but you're right about sticking with what most programs expect from a shell (darn you bash). I will open another PR with these!

**Comment by melund** [comment]
Couldn't you just start in interactive mode and then execute the lines in the script. Sort'a like a pasting everything into an interactive session.

**Comment by laerus** [comment]
yeah or `source` the script for more convenience,  i am used to python's `-i` flag when experimenting and i thought bash did the same but apparently it sets up interactive related things except entering the repl.

**Comment by scopatz** [comment]
Maybe we could use `-I` for this instead?

**Comment by laerus** [comment]
i'm okay with the current functionality, i think it can wait until more ppl need this feature.

**Pull request closed without merging**